[httpcheckreceiver] new receiver

codeboten opened the pull request:
The HTTP Check Receiver can be used for synthethic checks against HTTP endpoints. This receiver will make a request to the specified `endpoint` using the configured `method`.

Here's an example configuration:

```yaml
receivers:
  httpcheck:
    endpoint: http://endpoint:80
    method: GET
    collection_interval: 10s
```

The metrics being emitted:

```
ResourceMetrics #0
Resource SchemaURL:
ScopeMetrics #0
ScopeMetrics SchemaURL:
InstrumentationScope otelcol/httpcheckreceiver v0.62.0-6-g847c930f95
Metric #0
Descriptor:
     -> Name: httpcheck.duration
     -> Description: Measures the duration of the HTTP check.
     -> Unit: ms
     -> DataType: Gauge
NumberDataPoints #0
Data point attributes:
     -> http.url: Str(https://lightstep.com)
StartTimestamp: 2022-10-13 18:34:04.43338 +0000 UTC
Timestamp: 2022-10-13 18:34:14.436818 +0000 UTC
Value: 204
Metric #1
Descriptor:
     -> Name: httpcheck.status
     -> Description: 1 if the endpoint resulted in status http.status_code, otherwise 0.
     -> Unit: 1
     -> DataType: Sum
     -> IsMonotonic: false
     -> AggregationTemporality: Cumulative
NumberDataPoints #0
Data point attributes:
     -> http.method: Str(GET)
     -> http.status_class: Str(3xx)
     -> http.status_code: Int(200)
     -> http.url: Str(https://lightstep.com)
StartTimestamp: 2022-10-13 18:34:04.43338 +0000 UTC
Timestamp: 2022-10-13 18:34:14.436818 +0000 UTC
Value: 0
NumberDataPoints #1
Data point attributes:
     -> http.method: Str(GET)
     -> http.status_class: Str(4xx)
     -> http.status_code: Int(200)
     -> http.url: Str(https://lightstep.com)
StartTimestamp: 2022-10-13 18:34:04.43338 +0000 UTC
Timestamp: 2022-10-13 18:34:14.436818 +0000 UTC
Value: 0
NumberDataPoints #2
Data point attributes:
     -> http.method: Str(GET)
     -> http.status_class: Str(5xx)
     -> http.status_code: Int(200)
     -> http.url: Str(https://lightstep.com)
StartTimestamp: 2022-10-13 18:34:04.43338 +0000 UTC
Timestamp: 2022-10-13 18:34:14.436818 +0000 UTC
Value: 0
NumberDataPoints #3
Data point attributes:
     -> http.method: Str(GET)
     -> http.status_class: Str(1xx)
     -> http.status_code: Int(200)
     -> http.url: Str(https://lightstep.com)
StartTimestamp: 2022-10-13 18:34:04.43338 +0000 UTC
Timestamp: 2022-10-13 18:34:14.436818 +0000 UTC
Value: 0
NumberDataPoints #4
Data point attributes:
     -> http.method: Str(GET)
     -> http.status_class: Str(2xx)
     -> http.status_code: Int(200)
     -> http.url: Str(https://lightstep.com)
StartTimestamp: 2022-10-13 18:34:04.43338 +0000 UTC
Timestamp: 2022-10-13 18:34:14.436818 +0000 UTC
Value: 1
	{"kind": "exporter", "data_type": "metrics", "name": "logging"}
```

Fixes #10607

Commit:
[httpcheckreceiver] new receiver

The HTTP Check Receiver can be used for synthethic checks against HTTP endpoints. This receiver will make a request to the specified `endpoint` using the configured `method`.

Fixes #10607

mwear commented on receiver/httpcheckreceiver/scraper.go:
```diff
@@ -0,0 +1,83 @@
+func (h *httpcheckScraper) scrape(ctx context.Context) (pmetric.Metrics, error) {
+	now := pcommon.NewTimestampFromTime(time.Now())
+
+	if h.client == nil {
+		return pmetric.NewMetrics(), errClientNotInit
+	}
```
Any reason not to make this the first thing that happens in `scrape`? It will spare creating a timestamp in the event that the client isn't initialized.

codeboten replied:
Done

mwear commented on receiver/httpcheckreceiver/scraper.go:
```diff
@@ -0,0 +1,83 @@
+	success := 0
+	statusCode := -1
+	if err == nil {
+		if resp.StatusCode == http.StatusOK { // change this to use configurable status codes
+			success = 1
```
Is this comment suggesting that we'll be able to configure statuses other than 200 as successful? 👍 if so. This might be avoidable if we changed the metric representation to record the status.

codeboten replied:
Replied to this in my response to the other comment.

mwear commented on receiver/httpcheckreceiver/metadata.yaml:
```diff
@@ -0,0 +1,24 @@
+    type: string
+
+metrics:
+  httpcheck.success:
```
Is it worth considering any other representations of this metric? One thought is that if this were `httpcheck.status` where the value is the status code, then you wouldn't have to interpret what success means. Success, failure, etc could be determined on the query side.

codeboten replied:
That's correct, the representation could be the status code itself (which is currently an attribute). I don't have a strong preference one way or the other. The success/failure metric is similar to `probe_success` emitted by https://github.com/prometheus/blackbox_exporter. It does mean that this receiver will need to have a way for users to configure success/failure.

carlosalberto replied:
Out of curiosity, do you expect users to check the method attribute? e.g. probe both GET and POST in the same service?

codeboten replied:
i wouldn't expect someone to want to hit the same endpoint with different methods. i would expect consumers of the telemetry to want to be able to filter responses by the method though.

djaglowski replied:
I think it would make a lot of sense to keep the status code as an attribute, but also to change the meaning of the value from "success" to "a single observation of this status code". 

You would expect subsequent observations to look like this:
```
{ "url": "/foo", "method": "GET", "code": 200, "value": 1, "time": "t0" }
{ "url": "/foo", "method": "GET", "code": 200, "value": 1, "time": "t1" }
{ "url": "/foo", "method": "GET", "code": 404, "value": 1, "time": "t2" } // outage
{ "url": "/foo", "method": "GET", "code": 200, "value": 1, "time": "t3" }
```

See my other comment for more on why I think this model is useful for summarization.

jmacd replied:
See my suggestion to make this a 0-1 metric similar to or perhaps specified to create an OpenMetrics `StateSet`. @dashpole note the connection to #1712 and #2409.

dashpole replied:
I've reviewed your other comment, and I think this would be a great way to represent this sort of information!  A unit seems like the best place I can think of for putting that information, and it makes sense in contexts other than Prometheus.

dashpole replied:
Seems like we could take a similar approach for info metrics as well.

jmacd replied:
Sounds good!

carlosalberto commented on receiver/httpcheckreceiver/scraper.go:
```diff
@@ -0,0 +1,86 @@
+
+	now := pcommon.NewTimestampFromTime(time.Now())
+
+	u, err := url.Parse(h.cfg.Endpoint)
```
Out of curiosity: can this change during the lifetime of the receiver?

carlosalberto replied:
(the endpoint)

codeboten replied:
it shouldn't. i'll remove the code that parses it here.

carlosalberto commented:
Symbolic approval, this looks good! Also interested to help improve this with a few potential additions ;)

codeboten commented on receiver/httpcheckreceiver/metadata.yaml:
```diff
@@ -0,0 +1,32 @@
+  httpcheck.success:
+    description: The response of the endpoint resulted in status 200.
+    enabled: true
+    gauge:
```
Right, the way I was thinking about is that using a gauge, users would setup alerting based on individual checks failing rather than thinking about it in terms of aggregates. I can see how modeling this using a sum + delta aggregation would work as well, although this would mean that users couldn't infer meaning from the value being reported without looking at the attributes as well.

djaglowski replied:
I see what you're saying about it being difficult to alert on. What if the "status_code: 200" data point is always reported, either as 1 or 0? This would have no impact on aggregations, but also would allow for easy alerting on a specific check.

codeboten replied:
I guess then the question becomes why is `status_code: 200` special? Naming this `success` while combining it with a configuration option to define what success means to users i think would address most use cases.

Using a non-monotonic counter with delta temporality here would still allow users to aggregate data on the `code` to determine what kinds of issues are occurring across multiple endpoints. Imagining the following reports:

```
{ "time": "hour 0", "url": "/foo1", "method": "GET", "code": 200, "value": 1 }
{ "time": "hour 0", "url": "/foo2", "method": "GET", "code": 500, "value": 0 } // outage
{ "time": "hour 1", "url": "/foo1", "method": "GET", "code": 500, "value": 0 } // outage
{ "time": "hour 1", "url": "/foo2", "method": "GET", "code": 500, "value": 0 } // outage
{ "time": "hour 2", "url": "/foo1", "method": "GET", "code": 200, "value": 1 } // recovered
{ "time": "hour 2", "url": "/foo2", "method": "GET", "code": 200, "value": 1 } // recovered
```

djaglowski replied:
> I guess then the question becomes why is `status_code: 200` special?

I think @jmacd's [suggestion](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/14191#discussion_r987393250) resolves this, by treating none of the values as special and reporting 0's for all but one code. This is a bit verbose but is a well established format called [one-hot encoding](https://en.wikipedia.org/wiki/One-hot#Machine_learning_and_statistics).

codeboten commented on receiver/httpcheckreceiver/metadata.yaml:
```diff
@@ -0,0 +1,32 @@
+  httpcheck.duration:
+    description: Measures the duration of HTTP check.
+    enabled: true
+    sum:
+      aggregation: cumulative
+      monotonic: false
+      value_type: int
```
After thinking about this and looking through the specification, i *think* the right type to use here is a histogram. I've started looking at adding histogram support to mdatagen here: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/14691

jmacd replied:
I suggest Gauge, although Histogram is not wrong, just overkill. This touches on a lot of open specification issues!

jmacd commented on receiver/httpcheckreceiver/metadata.yaml:
```diff
@@ -0,0 +1,34 @@
+      monotonic: false
+    unit: 1
+    attributes: [http.url, http.status_code, http.method]
+  httpcheck.duration:
+    description: Measures the duration of HTTP check.
+    enabled: true
+    sum:
+      aggregation: cumulative
+      monotonic: false
+      value_type: int
+    unit: ms
+    attributes: [http.url]
```
```suggestion
  httpcheck.duration:
    description: Measures the duration of the HTTP check.
    enabled: true
    gauge:
      value_type: int
    unit: ms
    attributes: [http.url]
```

Suggestion is to make this a Gauge. Histogram feels like overkill. However, if we make this a Sum then the aggregation will not be meaningful.

There is an open issue discussing the (frequent) request for a synchronous Gauge instrument in the OpenTelemetry API. This kind of instrumentation is a perfect example of where this would be useful; when we get to that point, I expect that SDK views would be able to reconfigure a Gauge instrument as a Histogram instrument and vice versa -- as ways to increase or decrease the resolution of this metric. (Likely the Gauge histogram will fall out of of this; my goal is not to specify any of this until after OTel metrics reaches the 1.0 milestone.)

carlosalberto commented on receiver/httpcheckreceiver/metadata.yaml:
```diff
@@ -0,0 +1,32 @@
+
+metrics:
+  httpcheck.status:
+    description: 1 if the endpoint resulted in status http.status_code, otherwise 0.
```
200? 2xx?

codeboten replied:
updated the description to clarify that that 1 is if the check return a status code within a specific class

Commit:
add httpcheck.status_class attribute

Commit:
update readme

Commit:
adding metric for httpcheck errors

codeboten commented:
Thanks to everyone for all the input. I've updated `httpcheck.status` to produce a data point for each http response class (1xx, 2xx, 3xx, 4xx, 5xx). The resulting data looks something like this:
<img width="1661" alt="Screen Shot 2022-10-13 at 12 23 47 PM" src="https://user-images.githubusercontent.com/223565/195690901-cf4034fd-23c7-423b-8c77-9644135a6eff.png">

I've also added another metric to allow users to count the number of errors that occurred. Previously, i was using the status code `-1` which seemed a bit cryptic.

@mwear @djaglowski @jmacd PTAL

Commit:
ignore message as it may vary

Commit:
Merge branch 'main' into codeboten/httpcheck
